enhancements/cluster-logging/multi-cluster-log-forwarder.md

```diff
@@ -0,0 +1,217 @@
+
+The workflows make the following assumptions:
+
+* The **cluster-logging-operator** is deployed to the *openshift-logging* namespace
+* The **cluster-logging-operator** is able to watch any namespace
+
+### ClusterLogForwarder not NS:openshift-logging/Name:instance
+
+**cluster administrator** is a user:
+
+* responsible for maintaining the cluster
+* able to bind cluster roles to serviceaccounts  
+* that deploys the **cluster-logging-operator**
+
+**namespace administrator** is a user:
+
+* able to create a serviceaccount
+* manages a **ClusterLogForwarder** custome resource
+
+The general worklow:
+
+* The namespace administrator creates a service account to be used by a log collector
+* The cluster administrator binds cluster roles to the service account for the log types they are allowed to collect (e.g. audit, infrastructure)
+
+```
+    oc create clusterrolebinding kube-audit-log-collection --clusterrole=collect-audit-logs --serviceaccount=openshift-kube-apiserver:audit-collector
+```
+
+* The namespace administrator creates a **ClusterLogForwarder** CR with a name that matches the service account and the inputs for which they are allowed to collect
+
+```yaml
+    apiVersion: "logging.openshift.io/v1"
+    kind: ClusterLogForwarder
+    metadata:
+      name: audit-collector
+      namespace: openshift-kube-apiserver
+    spec:
+      pipelines:
+       - inputRefs: 
+         - audit
+         outputRefs:
+         - loki
+      outputs:
+      - name: loki
+        type: loki
+        url: https://mycenteralizedserver.some.place
+```
+
+* The namespace administrator creates a **ClusterLogging** CR with a name that matches the service account that only includes a spec for collection
+
+```yaml
+    apiVersion: "logging.openshift.io/v1"
+    kind: "ClusterLogging"
+    metadata:
+      name: audit-collector
+      namespace: openshift-kube-apiserver
+    spec:
+      collection:
+        type: "vector"
+```
+
+##### Verification and Validations 
+The operator willvalidate resources upon reconciliation of a **ClusterLogForwarder** and **ClusterLogging** CR.  Failure to meet any of the following conditions will stop the operator from deploying a collector and it will add error status to the resource.
+
+* The **ClusterLogForwarder** CR defines a valid spec
+* There exists a serviceaccount that is named the same as the **ClusterLogForwarder** CR
+* The serviceaccount is bound to clusterroles that allow the input spec of the **ClusterLogForwarder** CR
```

Comment:
This is tricky to grasp, can you elaborate a bit more on:
- Are the inputs some sort of virtual resources that can be listed in a ClusterRole?
- Is the serviceaccount w/o a binding to the ClusterRole not able to send logs? Or is the ClusterLogForwarder CR in some sort of degraded state? If latter, will the cluster-logging-operator do some sort of SAR to explictly check if the serviceacount has the ClusterRole?
- In addition does it need to be a ClusterRole? AFAIU the ClusterLogForwarder resource behaves in this case namespaced.

Reply:
> > Are the inputs some sort of virtual resources that can be listed in a ClusterRole?
> 
> The intention is for the operator manifest to include roles that map to our opinionated inputs. There will have to be documented listing of which roles are required in order to collect which input types. My expectation is we will define a convention tied to the name (e.g. collect-application-logs, collect-infrastructure-logs)
> 
> ...
> 
> Correct. The operator will perform a SAR for each source type that needs to be collected.

Both things should be documented here, the opinionated ClusterRoles for each input type as well as how the SAR looks against them. Effectively they are the core of this proposal regardless lifting naming/namespace limitations.

>  The CLF will be rejected outright if it asks for something not permitted.  We made the decision to no longer support a "degraded" state in order to be more clear regarding the actual state of collection and forwardering.
> 

How is this going to be achieved? ValidationWebhook? If yes this makes the webhook stateful because you need to talk to the API (i.e. doing SARs) and then accept or reject. Usually validation webhook are stateless and proof-check things that should never be valid inputs in the custom resource and cannot be expressed in the CRD's OpenAPI spec (e.g. two dependent fields).

Reply:
> Both things should be documented here, the opinionated ClusterRoles for each input type as well as how the SAR looks against them. Effectively they are the core of this proposal regardless lifting naming/namespace limitations.

I have pushed changed to make this more obvious but it was my impression this document does not include the explicit details that you are describing.  It looks more like this document is for getting enough of a gist to write JIRA cards and provide an implementation.  We have these details in the initial gdoc 

> How is this going to be achieved? ValidationWebhook? If yes this makes the webhook stateful because you need to talk to the API (i.e. doing SARs) and then accept or reject. Usually validation webhook are stateless and proof-check things that should never be valid inputs in the custom resource and cannot be expressed in the CRD's OpenAPI spec (e.g. two dependent fields).

We will not be using webhooks as we have been advised against their use but instead will make use of the existing validation model that lives in the operator and reports status.